subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,524 @@
+					LOG_DBG("tftp_get success - Server returned: %d",
+							extract_u16(tftpc_request_buffer + 2));
+
+					break;
+				}
+			}
+		}
+
+		/* No of times we had to re-tx this "request". */
+		no_of_retransmists++;
+
+		/* Log this out. */
+		LOG_DBG("tftp_send_request was either unable to get data \
+				 from the TFTP Server.");
+		LOG_DBG("Or failed to get any valid data.");
+		LOG_INF("no_of_retransmists = %d", no_of_retransmists);
+		LOG_INF("Are we re-transmitting: %s",
+				 (no_of_retransmists < TFTP_REQ_RETX) ?
+				 "yes" : "no");
+
+	} while (no_of_retransmists < TFTP_REQ_RETX);
+
+	/* Status? */
+	return server_response;
+}
+
+/* Name: tftp_connect
+ * Description: This function connects with the TFTP Server.
+ */
+static inline s8_t tftp_connect(s32_t sock, struct sockaddr_in *server)
+{
+	/* If the server information is not provided by the user, we
+	 * have to use the default server.
+	 */
+	if (server == NULL) {
+
+		/* Have to use the default server. Populate the IP address. */
+		inet_pton(AF_INET, TFTPC_DEF_SERVER_IP, &def_srv.sin_addr);
+
+		/* Update the pointer. */
+		server = &def_srv;
+	}
+
+	/* Connect with the TFTP Server. */
+	return connect(sock, (struct sockaddr *) server, sizeof(struct sockaddr_in));
+}
+
+/* Name: tftp_get
+ * Description: This function gets "file" from the remote server.
+ */
+int tftp_get(struct sockaddr_in *server, struct tftpc *client,
+			 const char *remote_file, const char *mode)
+{
+
+	s32_t   stat               = TFTPC_UNKNOWN_FAILURE;
+	s32_t   sock;
+	u16_t   server_response;
+	u8_t    no_of_retransmists = 0;
+
+	/* Re-init the global "block number" variable. */
+	tftpc_block_no = 1;
+	tftpc_index    = 0;
+
+	/* Create Socket for TFTP. Use IPv4 / UDP as L4 / L3
+	 * communication layers.
+	 */
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+
+	/* Valid sock descriptor? If not, best to tell the
+	 * caller about it.
+	 */
+	if (sock < 0) {
+		LOG_ERR("Failed to create UDP socket (IPv4): %d", errno);
+		return -errno;
+	}
+
+	/* Connect with the address.  */
+	stat = tftp_connect(sock, server);
+	if (stat < 0) {
+		LOG_ERR("Cannot connect to UDP remote (IPv4): %d", errno);
+		return -errno;
+	}
+
+	/* Obtain Global Lock before accessing critical resources. */
+	if (k_mutex_lock(&tftpc_lock, K_FOREVER) != 0) {
+		LOG_ERR("Failed to obtain TFTP Client Semaphore");
+		return -errno;
+	}
+
+	/* Send out the request to the TFTP Server. */
+	server_response = tftp_send_request(sock, RRQ_OPCODE,
+				 remote_file, mode);
+
+	do {
+		/* Did the server respond with data? */
+		if (server_response == DATA_OPCODE) {
+
+			/* Good News - TFTP Server responded with data. Lets talk
+			 * to the server and get all data.
+			 */
+			stat = tftpc_process_resp(sock, client);
+
+			/* Lets get more data if data was successful or we got
+			 * a duplicate packet.
+			 */
+			if ((stat == TFTPC_DATA_RX_SUCCESS) ||
+				(stat == TFTPC_DUPLICATE_DATA)) {
+
+recv:
```

Comment:
I'd re-think the logic here - a `goto` label within a loop is not an easy-to-follow solution.

Reply:
I've refactored the logic to make it simpler.. Let me know what you think about it..